engine/execution/ingestion/engine.go

```diff
@@ -180,6 +186,187 @@ func (e *Engine) process(originID flow.Identifier, event interface{}) error {
 	}
 }
 
+func (e *Engine) finalizedUnexecutedBlocks() ([]flow.Identifier, error) {
+	// get finalized height
+	final, err := e.state.Final().Head()
+	if err != nil {
+		return nil, fmt.Errorf("could not get finalized block: %w", err)
+	}
+
+	// find the first unexecuted and finalized block
+	// we iterate from the last finalized, check if it has been executed,
+	// if not, keep going to the lower height, until we find an executed
+	// block, and then the next height is the first unexecuted.
+	// if there is only one finalized, and it's executed (i.e. genesis),
+	// then the firstUnexecuted is a unfinalized block, which is ok,
+	// because the next loop will ensure it only iterate through finalized
+	// block.
+	lastExecuted := final.Height
+
+	for ; lastExecuted > 0; lastExecuted-- {
+		header, err := e.state.AtHeight(lastExecuted).Head()
+		if err != nil {
+			return nil, fmt.Errorf("could not get header at height: %v, %w", lastExecuted, err)
+		}
+
+		executed, err := state.IsBlockExecuted(e.unit.Ctx(), e.execState, header.ID())
+		if err != nil {
+			return nil, fmt.Errorf("could not check whether block is executed: %w", err)
+		}
+
+		if executed {
+			break
+		}
+	}
+
+	firstUnexecuted := lastExecuted + 1
+
+	e.log.Info().Msgf("last finalized and executed height: %v", lastExecuted)
+
+	unexecuted := make([]flow.Identifier, 0)
+
+	// starting from the first unexecuted block, go through each unexecuted and finalized block
+	// reload its block to execution queues
+	for height := firstUnexecuted; height <= final.Height; height++ {
+		header, err := e.state.AtHeight(height).Head()
```

Comment:
⚠️ 

I am not quite sure to which locking you refer with your statement: 
> I made a fix to lock the queue during the reloading

The link just brings me to a view of all the changes of `engine/execution/ingestion/engine_test.go `. Assuming you meant this:
https://github.com/onflow/flow-go/blob/b6a9788dcde9462bdb99bc40894a7cd11a8bfaca/engine/execution/ingestion/engine.go#L286-L290

I am not quite sure whether this completely avoids the edge case. If the follower is running, it might call into the Finalizer, which will update the protocol state. If this happens, your calls to `e.state.Final()` might return _different_ snapshots. 
During your reloading, you query `e.state.Final()` multiple times: 
https://github.com/onflow/flow-go/blob/b6a9788dcde9462bdb99bc40894a7cd11a8bfaca/engine/execution/ingestion/engine.go#L191
https://github.com/onflow/flow-go/blob/b6a9788dcde9462bdb99bc40894a7cd11a8bfaca/engine/execution/ingestion/engine.go#L245
I don't think there are any guarantees that both calls return the same block (?)  
* Lets assume the finalized block is at height 10 in your call to `finalizedUnexecutedBlocks`
https://github.com/onflow/flow-go/blob/b6a9788dcde9462bdb99bc40894a7cd11a8bfaca/engine/execution/ingestion/engine.go#L265-L273
* then, in your call to `pendingUnexecutedBlocks` it has increased to 11. 
* I think you will miss block 11 in this case, because `Pending()`  only returns the descendants of  block 11 but not block 11 itself.  

#### Suggestion

you could query the Finalized State at the beginning before this call:
https://github.com/onflow/flow-go/blob/b6a9788dcde9462bdb99bc40894a7cd11a8bfaca/engine/execution/ingestion/engine.go#L136
and pass the returned snapshot into your methods. A state snapshot is by construction immutable. Therefore, you would execute all your operations with one consistent finalized block

Reply:
The way it works is that, on restart the ingestion engine will reload blocks, and calls the [`mempool.Run`](https://github.com/onflow/flow-go/pull/73/files#diff-e5c0dea70cf688717060471583da13cc5269a6f661e955613ed7ef698354cc1aR297). This call will acquire the lock for reading and writing to the execution queues in the `mempool`. That means, any calls to `mempool.Run` will be blocked until the lock is released. 

And since I made the entire reloading in `mempool.Run`, it blocks other processes and make them wait for it to finish. So during the reloading, if finalization happens and calls the `BlockProcessable`, which calls the `mempool.Run`, this call will then be blocked, until the reloading is finished.